### ISSUE TITLE
Geostationary IR processing

### ISSUE THREAD

**Issue by jjguerrette** [JCSDA/ioda-converters#242]
Has anybody moved the processing of geostationary instruments (e.g., ABI, AHI) past the initial test stage?  The basic capability was demonstrated in UFO PR [#432](https://github.com/JCSDA/ufo/pull/432).  This needs to be extended to more dates, possibly including the 15 April - 15 May 2018 test period.

What steps are needed?

**Comment by jjguerrette** [comment]
For MMM purposes, the assimilated quantity needs to include:
 + superobbing due to low model resolution
 + cloud detection
   1. First for removal of cloudy pixels or groups of pixels
   2. then for calculating cloud fraction of super-obbed pixels to be used in all-sky DA experiments

IODA formatted files need to include both clear and cloudy pixels.  If there is a data flow that already includes all of these procedures and data (i.e. from BUFR), that would work.

If not, I will start prototyping from a Fortran-based code that was originally developed for WRFDA.  The goal will be to develop a standalone ioda-converter that starts from raw ABI NC files.  This may be ported to Python or could be written entirely in Python.  The original goal is for functionality, not speed, but Fortran may allow for easier parallelism across the ABI grid (5424 x 5424).

We are debating whether to super-ob, possibly with box-car averaging, within the converter or in UFO.  Some cloud detection tests require knowing the background BT and/or radiances and an approximate tropopause temperature.  Those quantities will be available in UFO.  Thus it would be preferable to super-ob in UFO after the cloud detection is complete.  The super-obing, cloud detection, and cloud fraction calculation are easiest to perform with a 2-D grid of data.

Is anybody working on super-ob algorithms in UFO?  @rhoneyager @ytremolet @shlyaeva 

@srherbener, you mentioned a while ago that IODA may eventually read 2-D arrays for geostationary instruments.  Has there been any movement on that front?

Our group's first step will be a quick and dirty solution in order to get started.  Just checking if anybody else has gone down this road yet.

**Comment by srherbener** [comment]
@jjguerrette, I have it on my list to do multi-dimension storage in the ObsSpace container. A small amount of progress has been made for this, but nothing substantial yet.

I'm currently working on doing performance testing of the ODC reader in IODA (feature/odc-reader branch). The purpose of this is to provide some input for the upcoming decision between netcdf and ODB file types for obs data. Once this is done (estimate 1 to 2 weeks) I want to start on the multi-dimension storage work.

You may want to stay with Fortran for the converter program. With the existing Python converters we've run into some severe performance issues, which are tolerable with small test files, but not acceptable for production sized data. Python would probably still be great for directing traffic and managing conversion flows, but it seems that the actual conversion of one file format to another is going to ultimately have to be a compiled executable.

Another aspect of the conversion process is that I would like all the converters to use a common interface to the IODA file format (whether netcdf or ODB) so that duplication of code can be avoided, easier to maintain, etc. The IodaIO class would be that interface, but it is currently lacking an API layer. (That's another project on the list.) I wouldn't worry about this too much for the initial quick and dirty solution, but it is something to consider. If you are interested in helping develop this common interface, that would be great!

**Comment by jjguerrette** [comment]
Based on our offline discussion, I will work on a prototype converter for ABI in Fortran with MPI capability.  The reader will come from code previously written for WRFDA.  It will be trivial to add super-obbing to that reader and to exercise a subset of cloud detection tests that do not require CRTM outputs.  I will borrow the NC writer from @hailingz's GNSSRO converter.  To the extent possible, I will try to modularize the existing writer so that both converters are using the same code.  This Fortran-based NC writer may or may not be useful in future developments of the IODA IO API. 
 The ABI converter will be driven by a python script that will feed it the upstream raw NC files in a loop over a set of requested dates.

Future steps beyond the prototype will involve bringing the super-obbing and all cloud detection tests into IODA and UFO.  Both of those require some communication of data relating to neighboring pixels.  There are significant developments needed in IODA/UFO to reach the full level of capabilities. At lest the prototype will be a first draft of the capabilities that are needed and a starting point for future work.  The missing cloud-detection tests In the first-stage prototype solution will cause some degradation to the obs product (nominally 10% of cloudy pixels will go undetected), but that is acceptable.

**Comment by jjguerrette** [comment]
Ok, one more update.  The cloud detection tests need information about surface type.  Thus some information from a model is needed.  Ideally UFO would provide that info.  However, due to other challenges of performing super-obbing and the remainder of cloud-detection in IODA/UFO, we have opted to use WRFDA as a pre-processing tool.  WRFDA will produce an intermediate NC file (analogous to the NCDIAG format for GSI) that will be readily converted to IODANC.  This will be a temporary pathway.  Then we can slowly migrate capabilities over from WRFDA.

The cloud detection needs to be performed on the full raw data set (before super-obbing).  So the first step in JEDI would be to add super-obbing.  Then cloud detection can follow.

**Comment by jjguerrette** [comment]
Thus the ideal JEDI workflow is blocked by https://github.com/JCSDA/ufo/issues/447.  Probably this issue needs to be broken into smaller pieces.

**Comment by jjguerrette** [comment]
I have a working prototype for a WRFDA radiance ncdiag converter in feature branch `feature/wrfda_rad_diag`.  So far only the obs converter is implemented.  I do not know if we will need a geovals and/or obsdiag converter similar to GSI.

I tested this for an ncdiag file containing super-obbed goes-16-abi observations and the output looks reasonable.  The input file is 56MB and the output file is 24MB because many quantities are discarded.  The execution on a login node takes approximately 10 seconds.  I would like to set up a test case, but maybe I should thin the data more in order to reduce the test duration and file size.

**Comment by srherbener** [comment]
Great news!

You would only need the geovals files if you plan to run ufo-bundle (ie, without a model) and want to test the obs operators, filters, etc. with the new geostationary obs data.

The test you describe is still too big for unit testing (ctest).  If you can take 100s of pixels from an image and get meaningful results, then it would make sense to create a fast running test for unit testing (ctest). Otherwise, don't worry about it for now and stick with running the bigger tests manually. We can put the longer running test into the tier 2 tests after we get that scheme in place (unit tests are tier 1 in this scheme).

**Comment by jjguerrette** [comment]
I am working on creating a new obs and geovals file, because I would like to add a test for the new Cloud_Fraction geoval in UFO (https://github.com/JCSDA/ufo/pull/787).  The easiest way for me to do this that I see (already done) is writing an hofx feedback file and geoval file from an hofx application test with mpas-jedi.  This circumvents the need to convert from an old dataset (GSI or WRFDA), as long as we trust UFO to provide a correct initial answer.

I have done this with the 120km mpas-jedi ctest with 6 vertical levels and the current standard amsua_n19 obs file.  Just to clean up the netcdf files, I wrote a python converter for converting hofx to obs files and output geovals to input geovals.  I do not think this "conversion" is strictly necessary.

- The hofx application adds an `@hofx` version of each `ObsValue` variable.  This will be the new UFO reference.  Then all of the `@Gsi*` quantities can be removed.

- I think output geovals files can be used directly as inputs to UFO ctests, but they are a bit messy to look at.  So the converter simplifies to using `nlevs` and `nlevsp1` as dimensions instead of a unique dimension name for each geoval (e.g., replace `mass_content_of_rain_in_atmosphere_layer_nval` with `nlevs` for `mass_content_of_rain_in_atmosphere_layer`).  Also the doubles are converted to floats to save space.  The GeoVaLs writer could be configured to write floats instead of doubles though.

Anyway, maybe something I'm doing here is wrong or needlessly complicated.  Does this sound reasonable?

**Comment by srherbener** [comment]
@jjguerrette, sorry about the slow response. I think what you are doing is reasonable. You may still want the WRFDA hofx results for something to compare with as a check on the UFO results. (I'm thinking WRFDA since that is where the obs data came from.)

I'm going to add more assignees so that they can help with this.

**Comment by jjguerrette** [comment]
When we use WRFDA as a converter and for cloud detection, we also use the GFS analysis as the background.  GFS does not have hydrometeors on their analysis output, meaning we can't simulate all-sky radiances.  So we would need to perform forecasts in WRF in order to initialize hydrometeors.  That adds a great expense to this preprocessing/conversion.

**Comment by srherbener** [comment]
I thought for the obs/geovals generated from GSI that a model forecast is run, but we should probably hear from others since I'm not an expert in this process. I agree that it would be good to avoid having to run a forecast.

**Comment by shlyaeva** [comment]
@jjguerrette: I think this is a good approach, and I agree with all your points on the clean-up of the output files.
The only other way to generate files that I can think of is to have some other system (GSI?) run with using Cloud Fraction and have it save the GeoVaLs. I do not know whether GSI or any other system uses this capability in CRTM. @emilyhcliu, @rhoneyager: do you know?

**Comment by emilyhcliu** [comment]
@shlyaeva @jjguerrette  Yes, the cloud fraction can be generated from the GSI.  The current operational GSI does not run with cloud fraction, but the cloud fraction is stored in the GFS forecast files and can be used in CRTM.  I do have the plan to create the GSI diagnostic files with all hydrometeors and cloud fraction included.  The goal is to port the capability of assimilation hydrometeor-affected radiances from GSI to JEDI.  Making the UFO simulates cloudy radiances using a full set of hydrometeors and cloud fraction is the first step.

**Comment by jjguerrette** [comment]
Thank you @emilyhcliu for that useful description of your plans.  When the [UFO PR](https://github.com/JCSDA/ufo/pull/787) for cloud fraction geovals is merged, UFO will be ready.  We are already using this code with MPAS successfully.

@shlyaeva, so either I can generate both geoval and reference hofx files using mpas-jedi+UFO or we can wait on some equivalent output from GSI.  Or both.  Do you have a preference?  It's not likely our group will spend the time to create an equivalent WRFDA reference hofx.

**Comment by jjguerrette** [comment]
I'm closing this, because the ioda-converters aspect was addressed in #249. 
https://github.com/JCSDA/ufo/issues/786 has two remaining items that cover the geovals/reference output issues for UFO.
